framework/Models/HybridModels/LogicalModel.py

```diff
@@ -0,0 +1,231 @@
+      @ In, samplerType, string, is the type of sampler that is calling to generate a new input
+      @ In, **kwargs, dict,  is a dictionary that contains the information coming from the sampler,
+           a mandatory key is the sampledVars'that contains a dictionary {'name variable':value}
+      @ Out, newInputs, dict, dict that returns the new inputs for each sub-model
+    """
+    self.raiseADebug("{}: Create new input.".format(self.name))
+    # TODO: standardize the way to handle code/external model/rom inputs
+    modelToRun = self.controlFunction.evaluate("evaluate", kwargs)
+    if modelToRun not in self.modelInstances:
+      self.raiseAnError(IOError, 'Model (i.e. {}) returned from "ControlFunction" is not valid!'.format(modelToRun),
+                        'Available models are: {}'.format(','.join(self.modelInstances.keys())))
+    kwargs['modelToRun'] = modelToRun
+    if self.modelInstances[modelToRun].type == 'Code':
+      codeInput = []
+      for elem in myInput:
+        if isinstance(elem, Files.File):
+          codeInput.append(copy.deepcopy(elem))
+      return (codeInput, samplerType, kwargs)
+    return (myInput, samplerType, kwargs)
+
+  def _externalRun(self, inRun, jobHandler):
+    """
+      Method that performs the actual run of the logical model (separated from run method for parallelization purposes)
+      @ In, inRun, tuple, tuple of Inputs (inRun[0] actual input, inRun[1] type of sampler,
+        inRun[2] dictionary that contains information coming from sampler)
+      @ In, jobHandler, instance, instance of jobHandler
+      @ Out, exportDict, dict, dict of results from this hybrid model
+    """
+    self.raiseADebug("{}: External Run".format(self.name))
+    originalInput = inRun[0]
+    samplerType = inRun[1]
+    inputKwargs = inRun[2]
+    identifier = inputKwargs.pop('prefix')
+    # TODO: execute control function, move this to createNewInput
+    modelToRun = inputKwargs.pop('modelToRun')
+    inputKwargs['prefix'] = modelToRun + utils.returnIdSeparator() + identifier
+    inputKwargs['uniqueHandler'] = self.name + identifier
+    moveOn = False
+    while not moveOn:
+      if jobHandler.availability() > 0:
+        self.modelInstances[modelToRun].submit(originalInput, samplerType, jobHandler, **inputKwargs)
+        self.raiseADebug("Job submitted for model", modelToRun, "with identifier", identifier)
+        moveOn = True
+      else:
+        time.sleep(self.sleepTime)
+    while not jobHandler.isThisJobFinished(inputKwargs['prefix']):
+      time.sleep(self.sleepTime)
+    self.raiseADebug("Job finished", modelToRun, "with identifier", identifier)
+    finishedRun = jobHandler.getFinished(jobIdentifier=inputKwargs['prefix'], uniqueHandler=inputKwargs['uniqueHandler'])
+    evaluation = finishedRun[0].getEvaluation()
+    if isinstance(evaluation, rerror):
+      self.raiseAnError(RuntimeError, "The model", modelToRun, "identified by", finishedRun[0].identifier, "failed!")
```

Comment:
I think we still need this. Since we submit the models through client, this check seems to me is more informative. It will indicate which sub-model is failed.